[optim] NAdam fold state updates into functional

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* #69981
* #69980
* #71335
* __->__ #71334
* #71333

Differential Revision: [D33767864](https://our.internmc.facebook.com/intern/diff/D33767864)

mikaylagawarecki committed:
[optim] NAdam fold state updates into functional

[ghstack-poisoned]

albanD reviewed (approved):
Removal of recomputation!

mikaylagawarecki committed:
Update on "[optim] NAdam fold state updates into functional"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "[optim] NAdam fold state updates into functional"


Differential Revision: [D33767864](https://our.internmc.facebook.com/intern/diff/D33767864)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767864).
